WebRegistry support for RunCommand and Run

andyk opened the pull request:
This implements support and tests for the RunCommand and Run to `WebRegistry` and simplifies the implementations of the http handling in `WebRegistry`.

See examples of how to use the newly supported API calls in `web/registry/tests/test_integration.py`.

This squashes existing and new Django migrations into a single migration file which means you need to delete and re-recreate your aos_web postgres database from scratch for it to work.

Also fixes typo in docs: replace incorrect instances of `agentos/utils.py`` with correct ``tests/utils.py``.

Remaining TODO in follow-on PRs:
* add tests for adding & getting AgentRuns
* Add support for backup_registries.

andyk commented:
@nickjalbert I think this is ready for a review!

nickjalbert commented on agentos/registry.py:
```diff
@@ -474,19 +436,47 @@ def add_run_artifacts(
+        for spec_key, spec_val in [
+            (k, v)
+            for k, v in flat_spec.items()
+            if k.endswith("_link") or not v
```
It might be useful to have a canonical set of keys for each spec type defined somewhere (maybe in `agentos/specs.py`?) and just drop any key not in the canonical set (and/or have some casting and validation functions built on top of the spec definitions).

andyk replied:
I've added classes in `agentos/specs.py` like `ComponentSpecKeys` which whose members are strings that represent the spec keys for each spec type. I'm going to leave this to a future PR which I've already started thinking about for #250

nickjalbert commented on agentos/run.py:
```diff
@@ -36,8 +41,19 @@ class Run:
+    output as input. In this case, the relationship is causal.
     """
 
+    # TODO: Decide if pointers to other runs should be special some how.
```
How are these pointers currently stored?  As params?

andyk replied:
MLflow uses tags for storing pointers to a Run's parent. The key for the tag to use is defined in `mlflow.utils.mlflow_tags.MLFLOW_PARENT_RUN_ID`. Actually that same file defines a bunch of standard MLflow tag keys.

I think for now, the only form of pointers to other runs that we manage are parents.

nickjalbert commented on web/registry/models.py:
```diff
@@ -145,6 +137,12 @@ def create_from_dict(component_spec_dict: Dict) -> List:
             # TODO - When we have accounts, we need to check the the user
             #        has permission to create a new version of this Component
             #        (i.e. if the name already exists but not the version).
+            print(Component.objects.all())
```
This print statement probably won't be great as the DB grows. :)

andyk replied:
Removed

nickjalbert commented on web/registry/models.py:
```diff
@@ -187,15 +185,12 @@ def _equals_spec(self, other_spec):
-            f"<Repo {self.pk}: {self.identifier} (type {self.type}) "
-            f"at {self.url}>"
-        )
+        return f"<Repo '{self.identifier}' type {self.type} " f"at {self.url}>"
```
Merge this into one f string?

andyk replied:
Done.

nickjalbert commented on web/registry/tests/test_registry.py:
```diff
@@ -1,64 +1,74 @@
-import yaml
-import json
+# import yaml
```
remove these commented out imports?

andyk replied:
done.

nickjalbert commented on web/registry/tests/test_registry.py:
```diff
@@ -1,64 +1,74 @@
-        response = self.client.post(url, data)
-        self.assertEqual(response.status_code, 200)
-        self.assertEqual(Run.objects.count(), 2)
+    # TODO: update these test once we've finished reworking ingest_spec
```
These TODOs are still pending in a follow on PR, right?

andyk replied:
Yep. Done.

nickjalbert commented on web/registry/tests/test_integration.py:
```diff
@@ -84,17 +82,39 @@ def test_web_registry_integration(self):
+        comp_run = simple_component.run("add_to_init_member", param_set)
+        run_cmd = comp_run.run_command
+        print("trying to add run_command_spec to web_registry:")
+        pprint.pprint(run_cmd.to_spec())
```
I'm not sure how you feel, but I have a vague intuition that we shouldn't leave informational print statements in tests (see prints below as well).  

Relatedly: we should probably clean up the informational prints a bit in the actual runtime at some point too. 🤪

andyk replied:
I've deleted a bunch and converted some into logging statements, mostly at the DEBUG level. I've held off removing all of the print statements in the core component creation process (the ones about adding dependencies) since i think they are currently kind of nice for demo purposes.

nickjalbert commented on web/registry/tests/test_integration.py:
```diff
@@ -84,17 +82,39 @@ def test_web_registry_integration(self):
+        run_cmd = comp_run.run_command
+        print("trying to add run_command_spec to web_registry:")
+        pprint.pprint(run_cmd.to_spec())
+        web_registry.add_run_command_spec(run_cmd.to_spec())
```
It might be worth making some asserts to confirm backend objects get created,  e.g.

```python
self.assertEqual(RunCommand.objects.count(), 0)
# do something that should create a RunCommand in the DB
self.assertEqual(RunCommand.objects.count(), 1)
```

andyk replied:
Done.

nickjalbert reviewed (approved):
LGTM!  Only minor comments from me; feel free to merge at will. :)

andyk commented:
@nickjalbert I just pushed another WIP commit to this branch which fixes some bugs that we found during my demo this past Tuesday. I'm not finished debugging and expanding our tests, and I haven't finished addressing your feedback above (e.g., removing informational print statements and removing some commented out cruft code) so this is not ready for merge yet.

andyk commented:
@nickjalbert I think this addresses your feedback and fixes the bugs that we hit during the demo in staff meeting on 2/15.

nickjalbert commented:
Excellent!  Merging.  Let's run through the demo when we meet today!
